Replace existing simulator config & Fix simulator

jack60612 opened the pull request:
Remove old simulator config from config file, and add new config to the file while adding backwards compatibility support and allowing simulator to be more easily customized
Changes were also made to tests that might decrease flakiness 
Selected_network (Implemented)
Introducer_peer (implemented)
Key_fingerprint: (Implemented)
Auto_farm: on/off (Implemented) & working well
Plot_directory (implemented)
Farming_address ( Implemented)


Tests were added

Commit:
redesign simulator config loader

Commit:
edit initial config file and use it in start_simulator

also final lint cleanup

Commit:
small change

Commit:
aarggggghhhh

LINTTTTTTTTTTTTTTTT

Commit:
fix lint for the last time

DAMN END OF LINE

mariano54 commented on chia/util/initial-config.yaml:
```diff
@@ -550,3 +546,27 @@ wallet:
+simulator:
+  # The full node simulator (if run) will run on this port
+  port: 8444
+
```
If the config variables are already in the full node section, there is no need to repeat them here

jack60612 replied:
These will be changed by the wizard later, and they allow the existing simulator to continue working

jack60612 replied:
The existing simulator which is designed to work with the gui expects the simulator to be at the default ports, so we will have the wizard change the ports to the testnet ports once it is run.

mariano54 replied:
can we have the GUI read the config instead? I think the way it works with existing GUI simulator is kind of bad because you can only run one at a time

jack60612 replied:
I removed the ports from it, should be good now

Commit:
Add Types & change port

Commit:
fix config loading for tests

Commit:
change config options

remove unnecessary options & change ports back

Commit:
Add additional config options for old simulator

jack60612 commented:
I will remove draft once i have a chance to test it

Commit:
Get working simulator & implement the requirements

i hope my code makes sense:)

Commit:
remove extra part of config

Commit:
fix config typos

Commit:
add callback for autofarm

we can call this to tell the node to farm a block

Commit:
add auto farm rpc to simulator

Commit:
add plot_directory to config

add separate plot dir to avoid issues where the plot dir is shared with tests and an assert fails

Commit:
remove port from config

no longer needed due to existing gui simulator being broken

Commit:
switch to dedicated plot creation function

Commit:
modify api for greater block customization

Added block type option & amount of block option

Commit:
disable auto farm for tests

breaks them all ;)
Also made sure to run autofarm while a transaction is in mempool if the mode was just changed

jack60612 commented:
All tested in working, this is part 1 of the new simulator

mariano54 commented on chia/simulator/SimulatorFullNodeRpcApi.py:
```diff
@@ -9,12 +9,30 @@
 
-    async def farm_tx_block(self, _request: Dict[str, object]) -> Dict[str, object]:
+    async def farm_block(self, _request: Dict[str, object]) -> Dict[str, object]:
         request_address = str(_request["address"])
```
I would instead provide an optional parameter: `guarantee_tx_block` which defaults to False. Also, FullBlock can be a tx block. We can never guarantee a non-tx block.  So if `guarantee_tx_block` is true, then we force a tx block to be created. Otherwise, we just create a block but we don't know whether it is tx or not.

mariano54 commented on chia/util/initial-config.yaml:
```diff
@@ -550,3 +546,27 @@ wallet:
+
+  selected_network: testnet0
+
+  # The Introducer is set to connect locally because the simulator is not on Mainnet
```
these variables are already present in full node config. We should remove it from here.

jack60612 replied:
Im removing selected_network and introducer, is that good?

mariano54 commented on chia/util/initial-config.yaml:
```diff
@@ -550,3 +546,27 @@ wallet:
+
+simulator:
+
+  selected_network: testnet0
```
this variable is already present for full node, so remove

mariano54 commented on chia/simulator/full_node_simulator.py:
```diff
@@ -1,26 +1,35 @@
             self.use_current_time = True
         else:
             self.use_current_time = False
+        if "auto_farm" in self.config["simulator"] and self.config["simulator"]["auto_farm"] is True:
```
you can do: 
```
auto_farm = self.config["simulator"].get("auto_farm", False)
```

mariano54 replied:
that's a default value in case it's not present. Same for config["simulator"]

jack60612 replied:
Yeah, i forgot about that. Thanks!

mariano54 commented on chia/simulator/full_node_simulator.py:
```diff
@@ -43,6 +52,26 @@ async def get_all_full_blocks(self) -> List[FullBlock]:
+            new_block = FarmNewBlockProtocol(self.bt.farmer_ph)
+            await self.farm_new_transaction_block(new_block)
+
+    async def update_autofarm_config(self, enable_autofarm: Optional[bool] = None) -> bool:
```
we should not pass in Optional, it's confusing. Better to pass in a bool, and check if None before calling this method

mariano54 reviewed (changes requested):
Can do the following to reduce plotting time:
* reduce k size
* reduce plot filter
* reduce number of plots to 2 or 3

The BlockTools changes are quite extensive...  but I guess they are necessary to use real plots and keys

mariano54 commented on tests/block_tools.py:
```diff
@@ -1394,8 +1421,10 @@ def get_challenges(
 
-def get_plot_dir() -> Path:
-    cache_path = DEFAULT_ROOT_PATH.parent.joinpath("test-plots")
+def get_plot_dir(plot_dir: Optional[str] = None) -> Path:
```
Should be Plot_dir: str="test-plots"

jack60612 replied:
What if plot_dir is None, this wont work

mariano54 replied:
It can't be none, we need test plots right?

jack60612 replied:
I meant what if the variable was left as default, which is none

mariano54 replied:
The default should be test-plots not None

jack60612 replied:
Ok so change it upstream too!

mariano54 replied:
Wait actually we already pass in test plots into init, so we can just save it as an instance variable and remove this one right?

jack60612 replied:
> Wait actually we already pass in plot dir into init, so we can just save it as an instance variable and remove this one right?

oh yeah lol, thats even better

jack60612 replied:
Oh wait, its not part of the block_tools class

mariano54 commented on tests/block_tools.py:
```diff
@@ -1405,8 +1434,8 @@ def get_plot_dir() -> Path:
 
-def get_plot_tmp_dir():
-    return get_plot_dir() / "tmp"
+def get_plot_tmp_dir(plot_dir: Optional[str] = None):
```
same here

mariano54 commented on tests/block_tools.py:
```diff
@@ -153,36 +155,49 @@ def __init__(
+                self._config[service]["selected_network"] = "testnet0"
+
+            # some tests start the daemon, make sure it's on a free port
+            self._config["daemon_port"] = find_available_listen_port("BlockTools daemon")
```
I think these configs can be passed in using `config_overrides` and this would make the code cleaner

jack60612 replied:
Ill move those too

jack60612 replied:
This can not be passed in through config_overrides, due to some of the changes being made on the base level of the config file

mariano54 replied:
Why can root level changes not work?

mariano54 commented on tests/block_tools.py:
```diff
@@ -153,36 +155,49 @@ def __init__(
-
-        with lock_config(self.root_path, "config.yaml"):
-            save_config(self.root_path, "config.yaml", self._config)
+        if automated_testing:
```
shouldn't we also override this for simulator?

jack60612 replied:
if we eliminate the extra cmds for the simulator, which we are, then yes

jack60612 replied:
so yes

Commit:
change farm_block for the better

Commit:
clean up autofarm

remove unnecessary optional

Commit:
update create block tools and allow additional plot creation tweaks

& do requested fixes

jack60612 commented:
> Can do the following to reduce plotting time:
> 
> * reduce k size
> * reduce plot filter
> * reduce number of plots to 2 or 3
> 
> The BlockTools changes are quite extensive... but I guess they are necessary to use real plots and keys

Blockfilter is already set to 1 by the test constants, the k size has been reduced to 19 and only 3 plots are made. This takes about 5-6 seconds and is not really noticeable

Commit:
change to k18's

Commit:
minor nitpick

mariano54 commented on chia/util/initial-config.yaml:
```diff
@@ -550,3 +546,18 @@ wallet:
+  auto_farm: True
+
+  # The key used by the simulator to farm blocks and send transactions
+  key_fingerprint:
```
Does this actually work, with no value?

jack60612 replied:
Yes it does, that's why I do the check for none in start_simulator

mariano54 commented on tests/block_tools.py:
```diff
@@ -257,25 +283,35 @@ def change_config(self, new_config: Dict):
 
-    async def setup_plots(self):
+    async def setup_plots(self, plots: Optional[int] = None, plot_size: int = 20):
+        if plots is not None:
```
Use the default argument value instead:

`ansyc def setup_plots(self, num_og_plots: int=15, num_pool_plots: int=5 ...)`

jack60612 replied:
I did that to try to avoid using another variable, however I'll just use self.automated_testing

jack60612 replied:
Done, thanks!

mariano54 commented:
Also, we might want to run a test that runs the simulator and requests to create a few blocks, to ensure that it doesn't break. We have no tests for these new API calls for example for autofarming.

Commit:
testing some changes

stop reading these descriptions, just ask me on keybase

Commit:
change plot directory name stuff

Commit:
add editing of root level keys

Commit:
use config_overrides

Commit:
update plot info

Commit:
re separate create_block_tools

mariano54 commented on chia/simulator/start_simulator.py:
```diff
@@ -22,54 +24,102 @@
     return kwargs
 
 
+def create_block_tools_simulator(
```
Having a new function here just adds a bunch of lines that we don't need. We can just create the BlockTools directly within main

mariano54 commented on tests/block_tools.py:
```diff
@@ -257,25 +287,33 @@ def change_config(self, new_config: Dict):
     def add_plot_directory(self, path: Path) -> None:
-        self._config = add_plot_directory(self.root_path, str(path))
+        try:
+            self._config = add_plot_directory(self.root_path, str(path))
```
Why do we need to catch the value error here?

jack60612 replied:
It will throw an error if the plot directory is already in the config, let me see if i can solve that better

mariano54 commented on tests/block_tools.py:
```diff
@@ -1394,8 +1433,8 @@ def get_challenges(
 
-def get_plot_dir() -> Path:
-    cache_path = DEFAULT_ROOT_PATH.parent.joinpath("test-plots")
+def get_plot_dir(plot_dir_name: str = "test-plots") -> Path:
```
We are hardcoding "test-plots" in multiple places, so I recommend not having a default value here

mariano54 commented on tests/block_tools.py:
```diff
@@ -1405,8 +1444,8 @@ def get_plot_dir() -> Path:
 
-def get_plot_tmp_dir():
-    return get_plot_dir() / "tmp"
+def get_plot_tmp_dir(plot_dir_name: str = "test-plots") -> Path:
```
same here

mariano54 commented on tests/block_tools.py:
```diff
@@ -2057,9 +2096,8 @@ def create_block_tools(
-    asyncio.get_event_loop().run_until_complete(bt.setup_keys())
-    asyncio.get_event_loop().run_until_complete(bt.setup_plots())
-
+    asyncio.run(bt.setup_keys())
```
let's revert this, I remember that this caused an issue before, but not exactly sure why.

Commit:
revert some changes as requested

change back get_plot_dir & undo change to asyncio.run

Commit:
replace unnessary catch while adding plot locations

switch to better if logic

Commit:
combine simulator_bt function into main

fix block_tools logic too

Commit:
fix lint

windows bad

mariano54 commented:
Tests are failing for some reason

Commit:
test auto_farming

Commit:
remove change that causes more harm then good

jack60612 commented:
> Tests are failing for some reason

i think it was caused by removing the "test-plots" default value.

jack60612 commented:
I added an auto_farming test

Commit:
move test over and unflake it

Commit:
make test do the bare minimum

Commit:
add proper block waiting or wait avoidance

Commit:
fix failing test

Commit:
add new rpc client & a small fix

Commit:
add new test & add some minor changes to start_sim to support it

New test tests whole user facing simulator with real keychain, gui and daemon.

jack60612 commented:
ci will fail for a little bit

Commit:
Merge branch 'main' into jn.simulator-config

Commit:
merge fixes

jack60612 commented:
ill change the user back once the changes are merged

Commit:
update block-cache version

CHANGE TO MY USER, REVERT THIS LATER

PLS REVERT BEFORE MERGE

fix version

Commit:
Revert "update block-cache version"

This reverts commit 3afbbc5f51df002b84af3ce8f0f48e91cb503225.

Commit:
block args from reaching simulator

Commit:
clean up test plot creation

Commit:
remove extra bt's being called

this reduces load time

Commit:
fix stupid bug

Commit:
damn lint

Commit:
Merge branch 'main' into jn.simulator-config

jack60612 commented:
OOPS

mariano54 commented on chia/simulator/SimulatorFullNodeRpcClient.py:
```diff
@@ -0,0 +1,21 @@
+        address = encode_puzzle_hash(target_ph, "txch")
+        await self.fetch("farm_block", {"address": address, "blocks": number_of_blocks, "guarantee_tx_block": True})
+
+    async def enable_auto_farming(self, enable_auto_farming: bool) -> bool:
```
We should probable rename to `set_auto_farming`, since if we turn it off, the word enable doesn't make sense.

mariano54 commented on chia/simulator/SimulatorFullNodeRpcClient.py:
```diff
@@ -0,0 +1,21 @@
+        assert result == enable_auto_farming
+        return bool(result)
+
+    async def get_auto_farming_status(self) -> bool:
```
If so, this one can be `get_auto_farming`.

mariano54 commented on chia/simulator/SimulatorFullNodeRpcClient.py:
```diff
@@ -0,0 +1,21 @@
+
+
+class SimulatorFullNodeRpcClient(FullNodeRpcClient):
+    async def farm_block(self, target_ph: bytes32, number_of_blocks: int = 1) -> None:
```
can we combine both of these into one RPC api?

jack60612 commented on chia/simulator/SimulatorFullNodeRpcApi.py:
```diff
@@ -9,12 +9,28 @@
+                await self.service.server.api.farm_new_block(req)
         return {}
+
+    async def auto_farm(self, _request: Dict[str, object]) -> EndpointResult:
```
separate this

Commit:
Revert "remove extra bt's being called"

This reverts commit f78debcc26c45a95bba6a369ddbd9e50c979c61d.

Commit:
Revert "clean up test plot creation"

This reverts commit d006a1a1963182531a97e69f258255d4ea2ce0be.

Commit:
align with 12204

Commit:
rename RPC's for clarity

Commit:
disable cache error when not running in automated mode

Commit:
update tests with new rpc's

oops i forgor

Commit:
finalize sim api

Commit:
Merge branch 'main' into jn.simulator-config

Commit:
something funky going on

Commit:
Revert "something funky going on"

This reverts commit afa1ddde27b3a45bd5d5ea4fabfd00e885266973.

Commit:
Merge branch 'main' into jn.simulator-config

Commit:
fix config

Commit:
add get_farming_ph rpc

Commit:
remove random simulation thing, why are we even using this.

mariano54 commented on tests/simulation/test_start_simulator.py:
```diff
@@ -0,0 +1,183 @@
+        assert await simulator_rpc_client.get_auto_farming()
+        await time_out_assert(10, simulator_rpc_client.set_auto_farming, False, False)
+        await simulator.autofarm_transaction(dummy_hash)  # this should do nothing
+        await asyncio.sleep(3)  # wait for block to be processed
```
Sleep is bad and causes flaky tests. You can do a timeout assert to wait for a specific block height

jack60612 replied:
It's not supposed to advance, the sleepyness is only for sanity checking. Ex: fail condition

mariano54 replied:
got it thanks

Commit:
Merge branch 'main' into jn.simulator-config

Commit:
update tests

Commit:
simplify test code & fix lint

Commit:
lol one of my own pr's broke another one of my pr's

git funny

wjblanke reviewed (approved):
aok
